text/0000-pin.md

```diff
@@ -0,0 +1,408 @@
+
+# Reference-level explanation
+
+## The `Unpin` auto trait
+
+This new auto trait is added to the `core::marker` and `std::marker` modules:
+
+```rust
+pub unsafe auto trait Unpin { }
+```
+
+A type that implements `Unpin` can be moved out of one of the pinned reference
+types discussed later in this RFC. Otherwise, they do not expose a safe API
+which allows you to move a value out of them. Because `Unpin` is an auto trait,
+most types in Rust implement `Unpin`. The types which don't are primarily
+self-referential types, like certain generators.
+
+This trait is a lang item, but only to generate negative impls for certain
+generators. Unlike previous `?Move` proposals, and unlike some traits like
+`Sized` and `Copy`, this trait does not impose any compiler-based semantics
+types that do or don't implement it. Instead, the semantics are entirely
+enforced through library APIs which use `Unpin` as a marker.
+
+## `Pin` and `PinMut`
+
+`Pin` and `PinMut` are two types added to `core::mem` and `std::mem`. They are
+analogous to `&T` and `&mut T` respectively.
+
+```rust
+#[fundamental]
+pub struct Pin<'a, T: ?Sized + 'a> {
+    data: &'a T,
+}
+
+#[fundamental]
+pub struct PinMut<'a, T: ?Sized + 'a> {
+    data: &'a mut T,
+}
+```
+
+### Safe APIs
+
+They both implement `Deref`, but `PinMut` only implements `DerefMut` if the
+type it references implements `Unpin`:
+
+```rust
+impl<'a, T: ?Sized> Deref for Pin<'a, T> { ... }
+
+impl<'a, T: ?Sized> Deref for PinMut<'a, T> { ... }
+
+impl<'a, T: Unpin + ?Sized> DerefMut for PinMut<'a, T> { ... }
+```
+
+They can only be safely constructed from references to types that implement
+`Unpin`:
+
+```rust
+impl<'a, T: Unpin + ?Sized> Pin<'a, T> {
+    pub fn new(reference: &'a T) -> Pin<'a, T> { ... }
+}
+
+impl<'a, T: Unpin + ?Sized> PinMut<'a, T> {
+    pub fn new(reference: &'a mut T) -> PinMut<'a, T> { ... }
+}
+```
+
+They also have a `borrow` function, which allows them to be transformed to pins
+with shorter lifetimes as may be necessary:
+
+```rust
+impl<'a, T: ?Sized> Pin<'a, T> {
+    pub fn borrow<'b>(this: &'b Pin<'a, T>) -> Pin<'b, T> { ... }
+}
+
+impl<'a, T: ?Sized> PinMut<'a, T> {
+    pub fn borrow<'b>(this: &'b PinMut<'a, T>) -> PinMut<'b, T> { ... }
+}
+```
+
+They may also implement additional APIs as is useful for type conversions, such
+as `AsRef`, `From`, and so on. They implement `CoerceUnsized` as necessary to
+make coercing them into trait objects possible.
+
+### Unsafe APIs
+
+Both types can be unsafely constructed from references to types which may not
+implement `Unpin`. Users who use these constructors must know that the type
+they are passing a reference to will never be moved again after the Pin is
+constructed.
```

Comment:
Types aren't moved, data is. Do you mean that the data behind the reference will never be moved again?

It may also be worth saying explicitly if moving again after `'a` has ended is okay or not. "never" makes it sound like it is not okay, but usually everything happening with references is limited in scope by their lifetime. If this really means "never", that means it is always unsafe to use this method with a reference that one did not create themselves (because in that case you have no control over what happens when the lifetime ends).

Reply:
Never means never - the only valid states for the data inside the pin from now until the program terminates are to be live at this memory address or to no longer be live. (You are right about the implication of that).

Reply:
Okay. It may be worth saying that explicitly, just to be sure :)